apps/vaporgui/TrackBall.cpp

```diff
@@ -368,11 +370,26 @@ bool Trackball::ReconstructCamera(
 
 // Set the quaternion and translation from a viewer frame
 // Also happens to construct modelview matrix, but we don't use its translation
-void Trackball::setFromFrame(
+int Trackball::setFromFrame(
 	const std::vector<double>& posvec, const std::vector<double>& dirvec, 
 	const std::vector<double>& upvec, const std::vector<double>& centerRot,
 	bool persp
 ) {
+    VAssert( posvec.size() == 3 );
+    VAssert( dirvec.size() == 3 );
+    VAssert( upvec.size() == 3 );
+    VAssert( centerRot.size() == 3 );
+
+    // Check for valid up vector and view direction
+    if ( upvec[0] == 0 && upvec[1] == 0 && upvec[2] == 0 ) {
+        MSG_ERR("Invalid up vector");
+        return -1;
+    }
+    if ( dirvec[0] == 0 && dirvec[1] == 0 && dirvec[2] == 0 ) {
```

Comment:
Since these are floats we may want to check if they are very small rather than if they are 0. You can use FLT_EPSILON for that.

Reply:
Correct me if I'm wrong, but the viewing and up vectors are only invalid if all 3 components are 0.  If the components are 0, 0, and FLT_EPSILON, then the vector would be valid.